File: backend/app/schemas/rman.py
from pydantic import BaseModel
from typing import Optional


class RmanBackupRequest(BaseModel):
    oracle_version: str = "19c"
    # Backup type
    backup_type: str = "full"  # full / incremental / archivelog
    incremental_level: int = 1  # 0 or 1
    # Scope
    scope: str = "database"  # database / tablespace / datafile / controlfile / spfile
    tablespace_name: str = ""
    datafile_path: str = ""
    # Format
    format_tag: str = ""
    format_path: str = "/backup/rman"
    format_pattern: str = "%d_%T_%U"
    # Channels
    channel_count: int = 1
    channel_type: str = "disk"  # disk / sbt
    # Compression
    compression: str = "NONE"  # NONE / LOW / MEDIUM / HIGH / BASIC
    # Encryption
    encryption: str = "NONE"  # NONE / TRANSPARENT / PASSWORD / DUAL
    encryption_algorithm: str = "AES128"  # AES128 / AES192 / AES256
    encryption_password: str = ""
    # Options
    section_size: str = ""  # e.g. "1G"
    skip_offline: bool = False
    skip_readonly: bool = False
    skip_inaccessible: bool = False
    delete_input: bool = False  # for archivelog backup
    # Retention
    retention_policy: str = ""  # e.g. "REDUNDANCY 3", "RECOVERY WINDOW OF 7 DAYS"
    keep_until_time: str = ""
    # Check
    validate: bool = False
    crosscheck: bool = False
    # Tag
    tag: str = ""
    # Connect info (for script generation)
    host: str = "localhost"
    port: int = 1521
    username: str = ""
    password: str = ""
    service_name: str = ""
    sid: str = ""
    connect_type: str = "service"
    oracle_home: str = "/u01/app/oracle/product/19.3.0/dbhome_1"
    # Script options
    include_housekeeping: bool = True


class RmanRestoreRequest(BaseModel):
    oracle_version: str = "19c"
    restore_type: str = "database"  # database / tablespace / controlfile / spfile / archivelog
    tablespace_name: str = ""
    # Point-in-time
    pitr_time: str = ""  # e.g. "2026-05-16 12:00:00"
    pitr_scn: str = ""
    # Until sequence
    until_sequence: str = ""
    until_thread: str = ""
    # From backup
    from_tag: str = ""
    # Options
    preview: bool = False
    validate: bool = False
    # Connect info
    host: str = "localhost"
    port: int = 1521
    username: str = ""
    password: str = ""
    service_name: str = ""
    sid: str = ""
    connect_type: str = "service"
    oracle_home: str = "/u01/app/oracle/product/19.3.0/dbhome_1"


class RmanConfigRequest(BaseModel):
    oracle_version: str = "19c"
    settings: dict = {}
    host: str = "localhost"
    oracle_home: str = "/u01/app/oracle/product/19.3.0/dbhome_1"


class RmanResponse(BaseModel):
    command: str
    script: str
    warnings: list[dict] = []
    recommendations: list[str] = []
    steps: list[str] = []

File: frontend/src/api/rman.ts
import api from './index'

export interface RmanBackupParams {
  backup_type: string
  scope: string
  incremental_level?: number
  tablespace_name?: string
  datafile_path?: string
  format_tag?: string
  format_path?: string
  format_pattern?: string
  channel_count?: number
  channel_type?: string
  compression?: string
  encryption?: string
  encryption_algorithm?: string
  encryption_password?: string
  section_size?: string
  skip_offline?: boolean
  skip_readonly?: boolean
  skip_inaccessible?: boolean
  delete_input?: boolean
  retention_policy?: string
  tag?: string
  oracle_home?: string
  include_housekeeping?: boolean
  crosscheck?: boolean
  validate?: boolean
}

export interface RmanRestoreParams {
  restore_type: string
  tablespace_name?: string
  pitr_time?: string
  pitr_scn?: string
  until_sequence?: string
  until_thread?: string
  from_tag?: string
  preview?: boolean
  validate?: boolean
  oracle_home?: string
}

export interface RmanConfigParams {
  settings: Record<string, string>
  oracle_home?: string
}

export function generateRmanBackup(data: RmanBackupParams) {
  return api.post('/rman/backup', data)
}

export function generateRmanRestore(data: RmanRestoreParams) {
  return api.post('/rman/restore', data)
}

export function generateRmanConfig(data: RmanConfigParams) {
  return api.post('/rman/config', data)
}
